src/rust/catnap/queue.rs

```diff
@@ -6,85 +6,220 @@
+                    })
+                },
+                Err(Fail { errno, cause: _ }) if errno == libc::EWOULDBLOCK || errno == libc::EAGAIN => {
+                    // Operation in progress. Check if cancelled.
+                    drop(socket);
+                    if let Err(e) = yielder.yield_once().await {
+                        break Err(e);
+                    }
+                },
+                Err(e) => break Err(e),
+            }
         }
     }
 
-    /// Get underlying POSIX file descriptor.
-    pub fn get_fd(&self) -> Option<RawFd> {
-        self.fd
+    pub fn set_connect(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().connect()
     }
 
-    /// Set underlying POSIX file descriptor.
-    pub fn set_fd(&mut self, fd: RawFd) {
-        self.fd = Some(fd);
+    pub async fn connect(&mut self, remote: SocketAddrV4, yielder: Yielder) -> Result<(), Fail> {
+        loop {
+            let mut socket = self.socket.borrow_mut();
+            match socket.try_connect(remote) {
+                Ok(r) => break Ok(r),
+                Err(Fail { errno, cause: _ }) if retry_errno(errno) => {
+                    // Operation in progress. Check if cancelled.
+                    drop(socket);
+                    if let Err(e) = yielder.yield_once().await {
+                        break Err(e);
+                    }
+                },
+                Err(e) => break Err(e),
+            }
+        }
+    }
+
+    pub fn close(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().close()?;
+        self.socket.borrow_mut().try_close()
+    }
+
+    pub fn set_close(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().close()
+    }
+
+    pub async fn async_close(&mut self, yielder: Yielder) -> Result<(), Fail> {
+        loop {
+            let mut socket: RefMut<Socket> = self.socket.borrow_mut();
+            match socket.try_close() {
+                Ok(()) => break Ok(()),
+                Err(Fail { errno, cause: _ }) if errno == libc::EWOULDBLOCK || errno == libc::EAGAIN => {
+                    // Operation in progress. Check if cancelled.
+                    drop(socket);
+                    if let Err(e) = yielder.yield_once().await {
+                        break Err(e);
+                    }
+                },
+                Err(e) => break Err(e),
+            }
+        }
     }
 
-    /// Sets the underlying socket.
-    pub fn set_socket(&mut self, socket: &Socket) {
-        self.socket = socket.clone();
+    pub async fn push(
+        &mut self,
+        buf: &mut DemiBuffer,
+        addr: Option<SocketAddrV4>,
+        yielder: Yielder,
+    ) -> Result<(), Fail> {
+        loop {
+            let socket: Ref<Socket> = self.socket.borrow();
+            match socket.try_push(buf, addr) {
+                Ok(()) => {
+                    if buf.len() > 0 {
```

Comment:
Inverting this condition would allow to reduce the indentation for the block.

Reply:
Not exactly sure what this means but I need to check the buf.len() after the try_push call because it modifies the buf.

Reply:
If you check for `if buf.len() == 0` and break Ok, then you don't need the first part in `if` or `else`.

I mean this:

```
if buf.len() == 0 {
    break Ok(());
}
// Operation in progress. Check if cancelled.
drop(socket);
if let Err(e) = yielder.yield_once().await {
        break Err(e);
}
```
It's not a blocker though.

Reply:
Ah got it